v2/dtos/requests/deviceservice_test.go

```diff
@@ -0,0 +1,127 @@
+	noAdminState.Service.OperatingState = ""
+	invalidAdminState := testAddDeviceService
+	invalidAdminState.Service.OperatingState = "invalid"
+	noBaseAddress := testAddDeviceService
+	noBaseAddress.Service.BaseAddress = ""
+	invalidBaseAddress := testAddDeviceService
+	invalidBaseAddress.Service.BaseAddress = "invalid"
+	tests := []struct {
+		name          string
+		DeviceService AddDeviceServiceRequest
+		expectError   bool
+	}{
+		{"valid AddDeviceServiceRequest", valid, false},
+		{"invalid AddDeviceServiceRequest, no Request Id", noReID, true},
+		{"invalid AddDeviceServiceRequest, no Name", noName, true},
+		{"invalid AddDeviceServiceRequest, no OperatingState", noOperatingState, true},
+		{"invalid AddDeviceServiceRequest, invalid OperatingState", invalidOperatingState, true},
+		{"invalid AddDeviceServiceRequest, no AdminState", noAdminState, true},
+		{"invalid AddDeviceServiceRequest, invalid AdminState", invalidAdminState, true},
+		{"invalid AddDeviceServiceRequest, no BaseAddress", noBaseAddress, true},
+		{"invalid AddDeviceServiceRequest, no BaseAddress", invalidBaseAddress, true},
+	}
+	for _, tt := range tests {
+		t.Run(tt.name, func(t *testing.T) {
+			err := tt.DeviceService.Validate()
+			assert.Equal(t, tt.expectError, err != nil, "Unexpected addDeviceServiceRequest validation result.", err)
+		})
+	}
+}
+
+func TestAddDeviceService_UnmarshalJSON(t *testing.T) {
+	valid := testAddDeviceService
+	resultTestBytes, _ := json.Marshal(testAddDeviceService)
+	type args struct {
+		data []byte
+	}
+	tests := []struct {
+		name             string
+		addDeviceService AddDeviceServiceRequest
+		args             args
+		wantErr          bool
+	}{
+		{"unmarshal AddDeviceServiceRequest with success", valid, args{resultTestBytes}, false},
+		{"unmarshal invalid AddDeviceServiceRequest, empty data", AddDeviceServiceRequest{}, args{[]byte{}}, true},
+		{"unmarshal invalid AddDeviceServiceRequest, string data", AddDeviceServiceRequest{}, args{[]byte("Invalid AddDeviceServiceRequest")}, true},
+	}
+	fmt.Println(string(resultTestBytes))
+	for _, tt := range tests {
+		t.Run(tt.name, func(t *testing.T) {
+			var expected = tt.addDeviceService
+			err := tt.addDeviceService.UnmarshalJSON(tt.args.data)
+			if tt.wantErr {
+				require.Error(t, err)
+			} else {
+				require.NoError(t, err)
+				assert.Equal(t, expected, tt.addDeviceService, "Unmarshal did not result in expected AddDeviceServiceRequest.")
+			}
+		})
+	}
+}
+
+func Test_AddDeviceServiceReqToDeviceServiceModels(t *testing.T) {
+	valid := []AddDeviceServiceRequest{testAddDeviceService}
+	expectedDeviceServiceModel := []models.DeviceService{{
+		Name:           TestDeviceServiceName,
+		BaseAddress:    TestBaseAddress,
+		OperatingState: models.Enabled,
+		Labels:         []string{"MODBUS", "TEMP"},
+		AdminState:     models.Locked,
+	}}
+	tests := []struct {
+		name              string
+		addDeviceServices []AddDeviceServiceRequest
+	}{
+		{"valid AddDeviceServiceRequest", valid},
```

Comment:
Do you plan to add invalid test case? If, not change to non-table driven to reduce complexity.

Reply:
Changed to non-table driven testing.